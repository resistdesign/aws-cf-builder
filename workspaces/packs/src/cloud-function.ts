import { AWS } from '@aws-cf-builder/types';
import { createResourcePack } from '@aws-cf-builder/utils';

export const PLACEHOLDER_FUNCTION_CODE: AWS.Lambda.Function.Code = {
  ZipFile:
    "module.exports = {handler: async () => ({\n            statusCode: 200,\n            headers: {'Content-Type': 'application/json'},\n            body: '\"You did it!\"'\n          })};\n",
};

export type AddCloudFunctionConfig = {
  id: string;
  code?: AWS.Lambda.Function.Code;
  environment?: AWS.Lambda.Function.Environment;
  handler?: any;
  runtime?:
    | 'dotnetcore1.0'
    | 'dotnetcore2.0'
    | 'dotnetcore2.1'
    | 'dotnetcore3.1'
    | 'go1.x'
    | 'java11'
    | 'java8'
    | 'java8.al2'
    | 'nodejs'
    | 'nodejs10.x'
    | 'nodejs12.x'
    | 'nodejs14.x'
    | 'nodejs4.3'
    | 'nodejs4.3-edge'
    | 'nodejs6.10'
    | 'nodejs8.10'
    | 'provided'
    | 'provided.al2'
    | 'python2.7'
    | 'python3.6'
    | 'python3.7'
    | 'python3.8'
    | 'python3.9'
    | 'ruby2.5'
    | 'ruby2.7';
  timeout?: any;
  policies?: AWS.IAM.Role.Policy[];
};

export const addCloudFunction = createResourcePack(
  ({
    id,
    code = PLACEHOLDER_FUNCTION_CODE,
    environment = {
      Variables: {
        NODE_ENV: 'production',
      },
    },
    handler = 'index.handler',
    runtime = 'nodejs12.x',
    timeout = 30,
    policies = [
      {
        PolicyName: 'lambda-parameter-store',
        PolicyDocument: {
          Statement: [
            // By default, the lambda can do anything.
            {
              Effect: 'Allow',
              Action: ['*'],
              Resource: '*',
            },
          ],
          Version: '2012-10-17T00:00:00.000Z',
        },
      },
    ],
  }: AddCloudFunctionConfig) => {
    return {
      Resources: {
        [`${id}Role`]: {
          Type: 'AWS::IAM::Role',
          Properties: {
            ManagedPolicyArns: ['arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole'],
            AssumeRolePolicyDocument: {
              Version: '2012-10-17',
              Statement: [
                {
                  Action: ['sts:AssumeRole'],
                  Effect: 'Allow',
                  Principal: {
                    Service: ['lambda.amazonaws.com'],
                  },
                },
              ],
            },
            Policies: policies,
          },
        },
        [`${id}`]: {
          Type: 'AWS::Lambda::Function',
          Properties: {
            Timeout: timeout,
            Code: code,
            Environment: environment,
            Handler: handler,
            Role: {
              'Fn::GetAtt': [`${id}Role`, 'Arn'],
            },
            Runtime: runtime,
          },
        },
      },
    };
  }
);
