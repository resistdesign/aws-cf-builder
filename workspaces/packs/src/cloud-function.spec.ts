import { addCloudFunction } from './cloud-function';

describe('Cloud Function', () => {
  describe('addCloudFunction', () => {
    test('should add a cloud function and role', () => {
      const cft = addCloudFunction(
        {
          id: 'Basic',
        },
        { AWSTemplateFormatVersion: '2010-09-09' }
      );

      expect(cft).toMatchSnapshot();
    });
  });
});
