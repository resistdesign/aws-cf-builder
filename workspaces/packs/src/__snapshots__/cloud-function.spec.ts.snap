// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Function addCloudFunction should add a cloud function and role 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Resources": Object {
    "Basic": Object {
      "Properties": Object {
        "Code": Object {
          "ZipFile": "module.exports = {handler: async () => ({
            statusCode: 200,
            headers: {'Content-Type': 'application/json'},
            body: '\\"You did it!\\"'
          })};
",
        },
        "Environment": Object {
          "Variables": Object {
            "NODE_ENV": "production",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "BasicRole",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BasicRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17T00:00:00.000Z",
            },
            "PolicyName": "lambda-parameter-store",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
